### PR TITLE
Refactor RNNs

### LINKED ISSUE #4105: Refactor NStepRNNs' links
All NStepRNNs have similar implementation. It is difficult to add new feature for them. We can merge it.

### PR DESCRIPTION
**Pull request by unnonouno**
All NStepRNNs contain similar codes. I refactored them to reduce redundant codes.
fix #4105

### CODE REVIEW

File: chainer/links/connection/n_step_rnn.py [hunk 1]
```diff
@@ -121,36 +120,59 @@ def __call__(self, hx, xs, **kwargs):
...
+        ws = [[getattr(w, 'w%d' % i) for i in range(self.n_weights)]
+              for w in self]
+        bs = [[getattr(w, 'b%d' % i) for i in range(self.n_weights)]
+              for w in self]
 
-        hy, trans_y = self.rnn(
-            self.n_layers, self.dropout, hx, ws, bs, trans_x,
-            activation=self.activation)
+        args = [self.n_layers, self.dropout] + hxs + [ws, bs, trans_x]
+        result = self.__class__.rnn(*args)
```
Comment (okuta) [inline chainer/links/connection/n_step_rnn.py]: Why do you use `__class__` ?

File: chainer/links/connection/n_step_rnn.py [hunk 2]
```diff
@@ -121,36 +120,59 @@ def __call__(self, hx, xs, **kwargs):
...
+                hx = self.init_hx(xs)
+            else:
+                hx = permutate.permutate(hx, indices_array, axis=1, inv=False)
+            hxs.append(hx)
 
         trans_x = transpose_sequence.transpose_sequence(xs)
 
-        ws = [[w.w0, w.w1] for w in self]
-        bs = [[w.b0, w.b1] for w in self]
+        ws = [[getattr(w, 'w%d' % i) for i in range(self.n_weights)]
```
Comment (okuta) [inline chainer/links/connection/n_step_rnn.py]: How about packing in `__init__`?

File: chainer/links/connection/n_step_rnn.py (hunk 1 above)
Comment (unnonouno) [reply chainer/links/connection/n_step_rnn.py]: I don't know why, but without it, it does not work. I'll try to fix it.

**Comment by unnonouno** [comment]
fixed

**Comment by okuta** [comment]
LGTM!

**Pull request merged**